### ISSUE TITLE
Incorrect warning `explicit lifetimes could be elided`

### ISSUE THREAD

**Issue by ikezedev** [rust-lang/rust-clippy#13811]
### Summary

I get false positive to remove a lifetime, when indeed that lifetime is used.

### Lint Name

needless_lifetimes

### Reproducer

I tried this code:

```rust
use derivative::Derivative;

struct TestContext;

#[derive(Derivative)]
#[derivative(Debug)]
struct SessionTest<'a> {
    #[derivative(Debug = "ignore")]
    context: &'a TestContext,
}
```

I saw this happen:

```
warning: the following explicit lifetimes could be elided: 'a
  --> iam-smoke-tests/tests/sessions.rs:16:20
   |
16 | struct SessionTest<'a> {
   |                    ^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
help: elide the lifetimes
   |
16 - struct SessionTest<'a> {
16 + struct SessionTest'_> {
```

I expected to see this happen:

### Version

```text
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.83.0
LLVM version: 19.1.1
```

**Comment by samueltardieu** [comment]
The suggested change is also syntactically incorrect: `struct SessionTest'_>`

**Comment by samueltardieu** [comment]
This is due to a bug in the `derivative` crate, which is no longer maintained. `derivative` generates code but [misuses `quote::quote_spanned` to do so](https://github.com/mcarton/rust-derivative/blob/master/src/debug.rs#L102). Using `quote` instead would fix the problem. This  prevents Clippy from distinguishing code generated from the derivative macro attribute from code written by the user. Only the latter should be linted, as indeed the code generated by the macro contains useless references but this should not be reported.

The `derivative` crate is no longer maintained and should no longer be used. See [this security advisory](https://rustsec.org/advisories/RUSTSEC-2024-0388.html for an alternative.
